Added AMD GPU Operator deployment (#216)
rancher/rke2-product-docs: versions/latest/modules/en/pages/add-ons/gpu_operators.adoc

====

[#_deploy_amd_operator_verify]
== Verify the installation

After a few minutes, perform the following checks:

. Check that the kernel module was loaded on the node.
+
[,sh]
----
lsmod | grep amdgpu
----

. Check that the GPU was added as an allocatable resource.
+
[,sh]
----
kubectl get nodes -o custom-columns="NAME:.metadata.name,GPU:.status.allocatable.amd\.com/gpu"
----
+
You should see `1` or more for multi-GPU nodes in the `GPU` column.

. Check that the node labeller applied GPU hardware labels.
+
[,sh]
----
kubectl get node -o jsonpath='{.items[0].metadata.labels.amd\.com/gpu\.driver-version}{"\n"}'
kubectl get node -o jsonpath='{.items[0].metadata.labels.amd\.com/gpu\.product-name}{"\n"}'
kubectl get node -o jsonpath='{.items[0].metadata.labels.kmm\.node\.kubernetes\.io/version-module\.kube-amd-gpu\.amdgpu-sles}{"\n"}'
----

[#_deploy_amd_operator_troublshoot]
== Troubleshooting

For troubleshooting, refer to the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/troubleshooting.html[AMD GPU Operator Troubleshooting documentation].

[#_deploy_amd_operator_uninstall]

> == Cleanup

Author: ```suggestion
== Cleanup
For steps to uninstall the AMD GPU Operator and its dependencies, refer to the link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/uninstallation/uninstallation.html[official documentaion].
```

====

[,bash]
----
echo PATH=$PATH >> /etc/default/rke2-server
----

[#_deploy_amd_operator]
== Deploy AMD operator on SLES

The link:https://github.com/ROCm/gpu-operator[AMD GPU Operator] allows administrators of Kubernetes clusters to manage AMD GPUs just like CPUs.

On SLES, the operator uses precompiled `amdgpu` driver images from the link:http://registry.suse.com[Trusted container images repository]
built against the codestream's GA kernel. KMM relocates the modules to the node's
running kernel using SUSE's stable kABI guarantee--no recompilation needed across
kernel updates within the same codestream. The resulting image is pushed to your
registry of choice.

[TIP]
.Driver images for SLES
====
The prebuild driver images for specific SLES versions are available from the following links:

* SLES 16: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles16[]
* SLES 15: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles15[]
====

[TIP]
.Official AMD documentation
====
Find the official documentation for precompiled drivers at the following links:

> * link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[]

Reviewer: ```suggestion
```

need cleanup, 2 copies here

[TIP]
.Official AMD documentation
====
For full installation instructions, DeviceConfig reference, and troubleshooting, refer to the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[AMD GPU Operator official documentation].
====

[#_deploy_amd_operator_prerequisites]
=== Prerequisites

* Verify that a link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/index.html#supported-hardware[supported AMD GPU] PCI device is present.
+
[,sh]
----
lspci -nn | grep -i "amd\|ati\|display\|vga"
----

* Verify that the RKE2 cluster is up and running.
+
[,sh]
----
kubectl get nodes
----

> * Install link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html?utm_source=chatgpt.com#install-cert-manager[cert-manager].

Reviewer: ```suggestion
* Install link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html#install-cert-manager[cert-manager].
```

link cleanup

====

[#_deploy_amd_operator_verify]
== Verify the installation

After a few minutes, perform the following checks:

. Check that the kernel module was loaded on the node.
+
[,sh]
----
lsmod | grep amdgpu
----

. Check that the GPU was added as an allocatable resource.
+
[,sh]
----
kubectl get nodes -o custom-columns="NAME:.metadata.name,GPU:.status.allocatable.amd\.com/gpu"
----
+
You should see `1` or more for multi-GPU nodes in the `GPU` column.

. Check that the node labeller applied GPU hardware labels.
+
[,sh]
----
kubectl get node -o jsonpath='{.items[0].metadata.labels.amd\.com/gpu\.driver-version}{"\n"}'
kubectl get node -o jsonpath='{.items[0].metadata.labels.amd\.com/gpu\.product-name}{"\n"}'
kubectl get node -o jsonpath='{.items[0].metadata.labels.kmm\.node\.kubernetes\.io/version-module\.kube-amd-gpu\.amdgpu-sles}{"\n"}'
----

[#_deploy_amd_operator_troublshoot]
== Troubleshooting

For troubleshooting, refer to the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/troubleshooting.html[AMD GPU Operator Troubleshooting documentation].

[#_deploy_amd_operator_uninstall]

> == Cleanup

Reviewer: Here, we need to link to the official docs documenting uninstall steps - https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html#uninstallation

[TIP]
.Driver images for SLES
====
The prebuild driver images for specific SLES versions are available from the following links:

* SLES 16: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles16[]
* SLES 15: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles15[]
====

[TIP]
.Official AMD documentation
====
Find the official documentation for precompiled drivers at the following links:

* link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[]
[TIP]
.Official AMD documentation
====
For full installation instructions, DeviceConfig reference, and troubleshooting, refer to the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[AMD GPU Operator official documentation].
====

[#_deploy_amd_operator_prerequisites]
=== Prerequisites

* Verify that a link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/index.html#supported-hardware[supported AMD GPU] PCI device is present.
+
[,sh]
----
lspci -nn | grep -i "amd\|ati\|display\|vga"
----

* Verify that the RKE2 cluster is up and running.
+
[,sh]
----
kubectl get nodes
----

> * Install link:https://documentation.suse.com/cloudnative/rancher-manager/v2.14/en/installation-and-upgrade/other-installation-methods/http-proxy/install-rancher.html#_install_cert_manager[cert-manager].

Author: ```suggestion
* Install link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html?utm_source=chatgpt.com#install-cert-manager[cert-manager].
```

----
echo PATH=$PATH >> /etc/default/rke2-server
----

[#_deploy_amd_operator]
== Deploy AMD operator on SLES

The AMD GPU Operator allows administrators of Kubernetes clusters to manage AMD GPUs just like CPUs.

On SLES, the operator uses precompiled `amdgpu` driver images from the link:http://registry.suse.com[Trusted container images repository]
built against the codestream's GA kernel. KMM relocates the modules to the node's
running kernel using SUSE's stable kABI guarantee--no recompilation needed across
kernel updates within the same codestream. The resulting image is pushed to your
registry of choice.

[TIP]
.Driver images for SLES
====
The prebuild driver images for specific SLES versions are available from the following links:
* SLES 16: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles16[]
* SLES 15: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles15[]

> ====

Author: ```suggestion
The prebuild driver images for specific SLES versions are available from the following links:

* SLES 16: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles16[]
* SLES 15: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles15[]
====
```

[TIP]
.Official AMD documentation
====
Find the official documentation for precompiled drivers at the following links:
* link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[]

> * link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/drivers/precompiled-driver.html[]

Author: ```suggestion
Find the official documentation for precompiled drivers at the following links:

* link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[]
* link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/drivers/precompiled-driver.html[]
```

----
echo PATH=$PATH >> /etc/default/rke2-server
----

[#_deploy_amd_operator]
== Deploy AMD operator on SLES

> The AMD GPU Operator allows administrators of Kubernetes clusters to manage AMD GPUs just like CPUs.

Reviewer: ```suggestion
The link:https://github.com/ROCm/gpu-operator[AMD GPU Operator] allows administrators of Kubernetes clusters to manage AMD GPUs just like CPUs.
```

On SLES, the operator uses precompiled `amdgpu` driver images from the link:http://registry.suse.com[Trusted container images repository]
built against the codestream's GA kernel. KMM relocates the modules to the node's
running kernel using SUSE's stable kABI guarantee--no recompilation needed across
kernel updates within the same codestream. The resulting image is pushed to your
registry of choice.

[TIP]
.Driver images for SLES
====
The prebuild driver images for specific SLES versions are available from the following links:

* SLES 16: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles16[]
* SLES 15: link:https://registry.suse.com/repositories/third-party-amd-amdgpu-driver-sles15[]
====

[TIP]
.Official AMD documentation
====
Find the official documentation for precompiled drivers at the following links:

* link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[]

> * link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/drivers/precompiled-driver.html[]

Reviewer: let's link back to the official AMD documentation itself for most part because they are going to be continuing to receive updates as and when new releases happen, and will reduce overhead of consistent maintenance on our side.

```suggestion
[TIP]
.Official AMD documentation
====
For full installation instructions, DeviceConfig reference, and troubleshooting, refer to the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[AMD GPU Operator official documentation].
```

====

[#_deploy_amd_operator_prerequisites]
=== Prerequisites

> * Verify that an AMD GPU PCI device is present.

Reviewer: ```suggestion
* Verify that a link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/index.html#supported-hardware[supported AMD GPU] PCI device is present.
```

+
[,sh]
----
lspci -nn | grep -i "amd\|ati\|display\|vga"
----

* Verify that the RKE2 cluster is up and running.
+
[,sh]
----
kubectl get nodes
----

* Install link:https://documentation.suse.com/cloudnative/rancher-manager/v2.14/en/installation-and-upgrade/other-installation-methods/http-proxy/install-rancher.html#_install_cert_manager[cert-manager].
It is required for webhook TLS certificates.

[#_deploy_amd_operator_installing]
=== Installing AMD GPU Operator

. Add the operator repository and update it.
+
[,sh]
----
helm repo add rocm https://rocm.github.io/gpu-operator
helm repo update
----

. Install the operator.
+
[,sh]
----
helm install amd-gpu-operator rocm/gpu-operator-charts \
  --namespace kube-amd-gpu \
  --create-namespace \
  --version v1.5.1 \
  --set crds.defaultCR.install=false
----

. Wait for the operator's controller-manager pod to become ready.
+
[,sh]
----
kubectl wait --for=condition=ready pod \
  -l control-plane=controller-manager \
  -n kube-amd-gpu --timeout=300s
----

. Verify that the operator pods are running successfully.
+
[,sh]
----
kubectl get pods -n kube-amd-gpu

> ----

Reviewer: ```suggestion
[#_deploy_amd_operator_installing]
=== Installing AMD GPU Operator

Once the prerequisites are met and RKE2 is running, install the AMD GPU Operator:

[,sh]
----
helm repo add rocm https://rocm.github.io/gpu-operator
helm repo update

helm install amd-gpu-operator rocm/gpu-operator-charts \
  --namespace kube-amd-gpu \
  --create-namespace \
  --version <gpu-operator-version> \
  --set crds.defaultCR.install=false
----

For the latest release version, see the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html[AMD GPU Operator Helm installation documentation].

[NOTE]
====
If using a local plain-HTTP registry for driver images during testing, configure
containerd before applying the `DeviceConfig`. See
link:https://documentation.suse.com/cloudnative/rke2/latest/en/install/containerd_registry_configuration.html[RKE2 containerd registry configuration].
====

```

+
[,sh]
----
kubectl get nodes -o custom-columns="NAME:.metadata.name,GPU:.status.allocatable.amd\.com/gpu"
----
+
You should see `1` or more for multi-GPU nodes in the `GPU` column.

. Check that the node labeller applied GPU hardware labels.
+
[,sh]
----
kubectl get node -o jsonpath='{.items[0].metadata.labels.amd\.com/gpu\.driver-version}{"\n"}'
kubectl get node -o jsonpath='{.items[0].metadata.labels.amd\.com/gpu\.product-name}{"\n"}'
kubectl get node -o jsonpath='{.items[0].metadata.labels.kmm\.node\.kubernetes\.io/version-module\.kube-amd-gpu\.amdgpu-sles}{"\n"}'
----

. Run a pod to verify the GPU can be scheduled and detected.
+
[,sh]
----
kubectl apply -f - <<'EOF'
apiVersion: v1
kind: Pod
metadata:
  name: amd-smi
  namespace: default
spec:
  restartPolicy: Never
  containers:
  - name: amd-smi
    image: docker.io/rocm/pytorch:latest
    command: ["/bin/bash"]
    args: ["-c", "amd-smi version && amd-smi monitor -ptum"]
    resources:
      limits:
        amd.com/gpu: 1
      requests:
        amd.com/gpu: 1
EOF

kubectl wait --for=condition=ready pod/amd-smi --timeout=300s
kubectl logs amd-smi
kubectl delete pod amd-smi

> ----

Reviewer: ```suggestion

```

we can remove this section.

AMD official docs cover it here - https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html#test-gpu-workload-deployment

let's point back to ^

> [#_deploy_amd_operator_troublshoot]

Reviewer 2: There's a typo 'troublshoot,' and it can become a URL, so maybe worth fixing?

== Troubleshooting

For troubleshooting issues with the AMD GPU operator, perform the following checks:

. View the operator controller-manager logs.
+
[,sh]
----
kubectl logs -n kube-amd-gpu -l control-plane=controller-manager --tail=50
----

. View the KMM build job logs (active while the image is being built).
+
[,sh]
----
kubectl logs -n kube-amd-gpu -l kmm.node.kubernetes.io/job-type=build --tail=100
----

. View the KMM worker logs to monitor module loading.
+
[,sh]
----
kubectl logs -n kube-amd-gpu \
  -l kmm.node.kubernetes.io/job-type=device-config --all-containers --tail=50
----

. Check the details and status of the DeviceConfig and Module resources.
+
[,sh]
----
kubectl describe deviceconfig amdgpu-sles -n kube-amd-gpu
kubectl describe module amdgpu-sles -n kube-amd-gpu
----

. Inspect kernel messages related to the AMD GPU modules.
+
[,sh]
----
dmesg | grep -i "amdgpu\|amdkcl\|amdttm" | tail -30
----

> For more details, refer to the link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/troubleshooting.html[AMD GPU Operator Troubleshooting documentation].

Reviewer: ```suggestion
[#_deploy_amd_operator_troublshoot]
== Troubleshooting

For troubleshooting, refer to the
link:https://instinct.docs.amd.com/projects/gpu-operator/en/latest/troubleshooting.html[AMD GPU Operator Troubleshooting documentation].
```

same, point back to the official docs here

[#_deploy_amd_operator_uninstall]
== Cleanup

To uninstall the AMD GPU operator and its associated components, perform the following steps:

. Delete the custom `DeviceConfig` resource.
+
[,sh]
----
kubectl delete deviceconfig amdgpu-sles -n kube-amd-gpu
----

. Uninstall the AMD GPU operator and cert-manager Helm charts.
+
[,sh]
----
helm uninstall amd-gpu-operator -n kube-amd-gpu
helm uninstall cert-manager -n cert-manager
----

. Delete the `kube-amd-gpu` and `cert-manager` namespaces.
+
[,sh]
----
kubectl delete namespace kube-amd-gpu cert-manager

> ----

Reviewer: ```suggestion
[#_deploy_amd_operator_uninstall]
== Cleanup
```

same, point back to the official docs here - https://instinct.docs.amd.com/projects/gpu-operator/en/latest/installation/kubernetes-helm.html#uninstallation
